!!! FEATURE: Add Afx template for SiteKickstarter

Wyndala opened the pull request:
!!! FEATURE: Add Afx template for SiteKickstarter

This is the first approach for making the SiteKickstarter more versatile. Now you can easily add a custom generator package and kickstart some sorts of sites with one package.

* Add possibility to use more than one template for the site kickstarter
* Add Afx template in addition to the fluid template
* Add the generator selection to the backend module for site packages
* Removed the old generic generator service (this is breaking)

This is a breaking change because it removes the old, Fluid-based generator service.
In case you want to use the Fluid based rendering you can use the dedicated package at https://github.com/queoGmbH/FluidSiteGenerator

Resolves: #3074

bwaidelich commented:
Awesome, thanks for this. I need to look into this more closely, but I already added it to the [6.0 project](https://github.com/orgs/neos/projects/24) so that we can hopefully integrate it with the next major

bwaidelich commented:
Update: I just realized this was missing a corresponding feature ticket. I created one (#3099), linked it and added _that_ to the project board :)

daniellienert commented on Neos.Neos/Classes/Controller/Module/Administration/SitesController.php:
```diff
@@ -217,23 +218,42 @@ public function newSiteAction(Site $site = null)
      * @return void
      */
-    public function createSitePackageAction($packageKey, $siteName)
+    public function createSitePackageAction($packageKey, $generatorClass, $siteName)
```
As this changes the interface already, please add argument and return types

daniellienert commented on Neos.SiteKickstarter/Classes/Annotation/SitePackageGenerator.php:
```diff
@@ -0,0 +1,30 @@
+<?php
```
```suggestion
<?php
declare(strict_types=1);
```

daniellienert commented on Neos.SiteKickstarter/Classes/Annotation/SitePackageGenerator.php:
```diff
@@ -0,0 +1,30 @@
+namespace Neos\SiteKickstarter\Annotation;
+
+/**
+ * Class SitePackageGenerator
```
Please remove that comments

daniellienert commented on Neos.SiteKickstarter/Classes/Command/KickstartCommandController.php:
```diff
@@ -1,20 +1,13 @@
 <?php
-namespace Neos\SiteKickstarter\Command;
 
-/*
```
Why is this comment removed?

daniellienert commented on Neos.SiteKickstarter/Classes/Command/KickstartCommandController.php:
```diff
@@ -1,20 +1,13 @@
 <?php
```
```suggestion
<?php
declare(strict_types=1);
```

daniellienert commented on Neos.SiteKickstarter/Classes/Generator/AbstractSitePackageGenerator.php:
```diff
@@ -0,0 +1,8 @@
+
+abstract class AbstractSitePackageGenerator extends \Neos\Kickstarter\Service\GeneratorService
+{
+    abstract public function generateSitePackage($packageKey, $siteName);
```
Please add argument and return types

daniellienert commented on Neos.SiteKickstarter/Classes/Generator/AfxTemplateGenerator.php:
```diff
@@ -0,0 +1,205 @@
+<?php
```
```suggestion
<?php
declare(strict_types=1);
```

daniellienert commented on Neos.SiteKickstarter/Classes/Generator/AfxTemplateGenerator.php:
```diff
@@ -0,0 +1,205 @@
+            'type' => 'neos-site',
+            "require" => [
+                "neos/neos" => "*",
+                "neos/nodetypes" => "*"
```
The use of the nodetypes package is depreacated. We should not require it.

daniellienert commented on Neos.SiteKickstarter/Classes/Generator/AfxTemplateGenerator.php:
```diff
@@ -0,0 +1,205 @@
+     * @param string $siteName
+     * @return void
+     */
+    protected function generateSitesXml($packageKey, $siteName)
```
Please add proper typing throughout the classes

daniellienert commented on Neos.SiteKickstarter/Classes/Generator/FluidTemplateGenerator.php:
```diff
@@ -1,26 +1,21 @@
 <?php
-namespace Neos\SiteKickstarter\Service;
 
-/*
```
Please keep the license comment and add it also to the newly created files

daniellienert commented on Neos.SiteKickstarter/Classes/Generator/AfxTemplateGenerator.php:
```diff
@@ -0,0 +1,205 @@
+     * @param array $contextVariables
+     * @return string
+     */
+    protected function renderSimpleTemplate($templatePathAndFilename, array $contextVariables)
```
Looks like a method that should be put to an own service / utility class

daniellienert commented on Neos.Neos/Classes/Controller/Module/Administration/SitesController.php:
```diff
@@ -217,23 +218,42 @@ public function newSiteAction(Site $site = null)
+        $generatorServices = [];
+
+        if ($generatorServiceIsAvailable) {
+            $generatorClasses = $this->reflectionService->getAllSubClassNamesForClass(AbstractSitePackageGenerator::class);
```
IMHO the better way is to define a `SiteGeneratorInterface` and select classes implementing this interface

daniellienert commented on Neos.SiteKickstarter/Classes/Service/SitePackageGeneratorNameService.php:
```diff
@@ -0,0 +1,37 @@
+    /**
+     * @param string $generatorClass fully qualified namespace
+     */
+    public function getNameOfSitePackageGenerator($generatorClass)
```
Honestly I don't like the usage of special annotations here. Wouldn't it be more clear to add `getGeneratorName` / `getGeneratorDescription` to the generatorInterface or use just the className.

daniellienert commented on Neos.SiteKickstarter/Classes/Generator/FluidTemplateGenerator.php:
```diff
@@ -1,26 +1,21 @@
+ * @SiteKickstarter\SitePackageGenerator("Fluid Basic")
  */
-class GeneratorService extends \Neos\Kickstarter\Service\GeneratorService
+class FluidTemplateGenerator extends AbstractSitePackageGenerator
```
Should we really offer a Fluid template generator? I guess when asked, we recommend not to use fluid for site templating anymore.

daniellienert reviewed (changes requested):
Hey @Wyndala,
thanks a lot for your work! I added some comments - lots of nitpicks and some serious complains:
- I would not recommend to use annotations just for a display name, there are more straight-forward approaches. (but I am open to discuss that)
- I would like to discuss if we really should offer a fluid generator, we do recommend to not use fluid anymore

Cheers, 
Daniel

bwaidelich commented:
Great suggestions by Daniel and I agree to his comments

Wyndala commented:
Thx for the feedback -> I try to put it into the pr in this week :).

bwaidelich commented:
Update: I had created an issue (#3099) for that since the PR wasn't linked to a ticket. But in fact there _was_ an issue already (#3074) so I updated the description accordingly.

@Wyndala any news on this one?

jonnitto commented:
@Wyndala & @ahaeslich The feature freeze will be on the 9th of November, so it would be really awesome if you could finish this PR. If you have questions, don't wait to ask 😄

Wyndala commented:
@jonnitto I am currently on it :). I am currently testing my stuff. I try my best to get it work on the weekend.

Wyndala commented:
Hey @jonnitto, @daniellienert, @bwaidelich,

I just finished adding all the feedback.

I extracted the FluidTemplateGenerator to https://github.com/queoGmbH/FluidSiteGenerator . So in case somebody needs something like this, the code is not lost (even I don't like fluid and it is deprecated).

I tested the stuff in my biggest NEOS Installation and it worked fine. If there is something else I can do, just leave feedback again. After month of hard work my week looks a little bit better the next months. :)

Have a nice sunday,
Ralf

jonnitto commented on Neos.SiteKickstarter/Resources/Private/AfxGenerator/Fusion/Document/AbstractPage.fusion:
```diff
@@ -0,0 +1,58 @@
+    }
+
+    # Script includes before the closing body tag should go here
+		javascripts = Neos.Fusion:Array
```
```suggestion
    javascripts = Neos.Fusion:Join
```
Make usage of `Neos.Fusion:Join` and indent with spaces

jonnitto commented on Neos.SiteKickstarter/Resources/Private/AfxGenerator/Fusion/Document/AbstractPage.fusion:
```diff
@@ -0,0 +1,58 @@
+    # Script includes before the closing body tag should go here
+		javascripts = Neos.Fusion:Array
+    # This processor appends the rendered javascripts Array to the rendered template
+		@process.appendJavaScripts = ${value + this.javascripts}
```
```suggestion
    @process.appendJavaScripts = ${value + this.javascripts}
```
Indent with spaces

jonnitto commented on Neos.SiteKickstarter/Resources/Private/AfxGenerator/Fusion/Document/AbstractPage.fusion:
```diff
@@ -0,0 +1,58 @@
+		@process.appendJavaScripts = ${value + this.javascripts}
+  }
+
+  bodyJavaScripts = Neos.Fusion:Array
```
```suggestion
  bodyJavaScripts = Neos.Fusion:Join
```

jonnitto commented on Neos.SiteKickstarter/Resources/Private/AfxGenerator/Fusion/Document/AbstractPage.fusion:
```diff
@@ -0,0 +1,58 @@
+    }
+
+    # Add stylesheets to your page
+    stylesheets.site = Neos.Fusion:Tag {
+      tagName = 'link'
+      attributes.rel = 'stylesheet'
+      attributes.href = Neos.Fusion:ResourceUri {
+        # Change to your file path
+        path = 'resource://{packageKey}/Public/site.css'
+      }
+    }
```
```suggestion
     stylesheets.site = afx`
        <link rel="stylesheet" href={StaticResource.uri('{packageKey}', 'Public/site.css')} />
    `
```
Since we have now an [eel helper for static sources](https://github.com/neos/neos-development-collection/pull/3134), we could make use of this.

jonnitto commented on Neos.SiteKickstarter/Resources/Private/AfxGenerator/Fusion/Document/AbstractPage.fusion:
```diff
@@ -0,0 +1,58 @@
+    }
+
+    # Add header javascripts to your page
+    javascripts.site = Neos.Fusion:Tag {
+      tagName = 'script'
+      attributes.type = 'text/javascript'
+      attributes.src = Neos.Fusion:ResourceUri {
+        # Change to your file path
+        path = 'resource://{packageKey}/Public/site.js'
+      }
+    }
```
```suggestion
    javascripts.site = afx`
        <link type="text/javascript" defer href={StaticResource.uri('{packageKey}', 'Public/site.js')} />
    `
```
We could also use here the new Eel helper. And I suggest also to add the defer attribute…

jonnitto reviewed (changes requested):
I tested your PR and it works fine! ❤️ 
I made a few suggestions about the generated fusion code

bwaidelich reviewed (commented):
The change set still contains two empty Doctrine Migrations (Neos.SiteKickstarter/Migrations/Code/Version20161125002311.php and Neos.SiteKickstarter/Migrations/Code/Version20161125095901.php) that break the CI

jonnitto commented:
> The change set still contains two empty Doctrine Migrations (Neos.SiteKickstarter/Migrations/Code/Version20161125002311.php and Neos.SiteKickstarter/Migrations/Code/Version20161125095901.php) that break the CI

@bwaidelich Are you sure? I checked out this PR and these files have content. As I see, the filemode changed...

Wyndala commented:
Hey guys,

I fixed the weird file issue and now the ci is green. :)

I added jon's feedback and I think we are ready to merge (from my point of view) :)

daniellienert reviewed (commented):
Hey @Wyndala, looks much better now. Thanks a lot. 

I wasn't able to get it to work though, I get the error 

```
./flow kickstart:site Foo.Bar barfoo
Notice: Trying to access array offset on value of type null in /Users/ry25/Workspace/Projects/neosdev/Packages/Framework/Neos.Flow/Classes/Package/PackageManager.php line 373
```
Not sure if its related to your changes. @jonnitto have you tested it, does it work for you?

jonnitto commented:
> flow kickstart:site Foo.Bar barfoo

This works for me, yes. Tested it also without arguments.

Wyndala commented:
> > flow kickstart:site Foo.Bar barfoo
> 
> This works for me, yes. Tested it also without arguments.

I also tested it again on a fresh installation and it worked without any error

bwaidelich commented:
@daniellienert any reason still to block this one from your side?

daniellienert commented:
@bwaidelich, there are still parts of the code which would say is a bit over-engineered / bulky which I at least would discuss if there was time to do so, like the `SiteGeneratorCollectingService` or the `SitePackageGeneratorNameService`. But if you think its fine, Its fine for me.

bwaidelich commented:
Thanks for the feedback

> But if you think its fine, Its fine for me.

I didn't really get into it very deeply. In general I think it's a great improvement feature-wise and it's not userland code anyways but "just" a tool for devs. So I wouldn't be too concerned about architecture in that part. Also we still have ~3 weeks to fix potential bugs.

Having said that: Since I didn't really test/evaluate this, I'll abstain from voting  and leave it up to @jonnitto to decide

jonnitto commented:
I can't judge over the PHP part as @daniellienert, but as we say it is best practice to use `afx`, I think it is a really great feature to get rid of the old style.
